src/cache-prewarm.js

```diff
@@ -0,0 +1,81 @@
+import isUrl from "is-url";
+import pLimit from "p-limit";
+import { getCatalogs, getMatchForFilename } from "./catalogs.js";
+import { getFromUrlOrFile } from "./io.js";
+
+const limit = pLimit(10);
+
+function getFromUrlOrFileWithLimit(url, cache) {
+  return limit(() => getFromUrlOrFile(url, cache));
+}
+
+function normalizeUrl(ref) {
+  try {
+    const url = new URL(ref);
+    url.hash = "";
+    return url.toString();
+  } catch {
+    return null;
+  }
+}
+
+function getRemoteRefs(node) {
+  let refs = [];
+  if (Array.isArray(node)) {
+    for (const v of node) {
+      if (typeof v === "object" && v !== null) {
+        refs = refs.concat(getRemoteRefs(v));
+      }
+    }
+  } else if (typeof node === "object" && node !== null) {
+    for (const [k, v] of Object.entries(node)) {
+      if (k === "$ref" && typeof v === "string") {
+        const resolved = normalizeUrl(v);
+        if (resolved && isUrl(resolved)) {
+          refs.push(resolved);
+        }
+      }
+      if (typeof v === "object" && v !== null) {
+        refs = refs.concat(getRemoteRefs(v));
+      }
+    }
+  }
+  return Array.from(new Set(refs));
+}
+
+async function fetchAndRecurse(url, cache) {
+  const schema = await getFromUrlOrFileWithLimit(url, cache);
+  const refs = getRemoteRefs(schema);
+  await Promise.all(refs.map((ref) => fetchAndRecurse(ref, cache)));
+}
+
+async function prewarmSchemaCache(filenames, config, cache) {
+  const catalogs = getCatalogs(config);
+  const schemaLocations = new Set();
+
+  for (const filename of filenames) {
+    let catalogMatch;
+    try {
+      catalogMatch = config.schema
+        ? {}
+        : await getMatchForFilename(catalogs, filename, "debug", cache);
+    } catch {
+      catalogMatch = {};
+    }
+    const schemaLocation = config.schema || catalogMatch.location;
+
+    if (schemaLocation) {
+      schemaLocations.add(schemaLocation);
+    }
+    cache.resetCounters();
+  }
+
+  await Promise.all(
+    Array.from(schemaLocations)
+      .filter((schemaLocation) => isUrl(schemaLocation))
+      .map((url) => fetchAndRecurse(url, cache)),
+  );
+  cache.resetCounters();
```

Comment:
This.. kinda works. Tbh, it is probably fine for most cases due to URL normalisation and de-duplication. It all falls apart if we have multiple schemas that all have external `$refs` out to the same schemas though. There are two problems here:

1. Fundamentally, `this.callCounter[url]++;` is not a safe operation if we've got multiple promises concurrently modifying the same variable.
2. Even if this were a safe operation, this is not the right count because we're counting the number of calls across all promises, not the number of calls to a given URL internally within each schema resolution.